# Function 1

## 학습목표
 1. 함수의 이해
 2. 함수 구현 및 사용 숙지
 3. 재귀함수 구현 및 사용 숙지

## Function basic 

* 지금까지 무심코 코드에서 사용된 많은 함수들이 존재 합니다.
* 예를들면, sum, len, range 같은 함수 등이 있죠.
* 함수란 우리가 알고있는 개념처럼 주어진 입력(input)에 대해서 의도된 출력(output)를 전달하는 역할을 합니다.

* 그렇다면, 하나씩 살펴보겠습니다.
  > range 함수는 정수를 입력으로 전달하면 (0, 정수) 로 이루어진 리스트를 생성하는 역할을 합니다.
  
  > sum 함수는 리스트, 튜플등을 입력으로 전달하면 전체 아이템의 합을 출력으로 전달하는 역할을 합니다.
  
  > len 함수는 리스트, 튜플등을 입력으로 전달하면 아이템의 개수를 출력으로 전달하는 역할을 합니다.
  
* 그리고, 위의 함수들은 모두 python 내부에 이미 정의(구현)이 되어 있습니다.
* 위와 같은 함수를 내장함수(built-in function)이라고 합니다.

In [ ]:
# 내장 함수의 예

nums = list(range(11))
nums

print(sum(nums))
print(len(nums))


In [12]:
## 평균구하기 



sum = 0
for i in range (1,101):
    sum += i 

print(sum)    




5050


In [17]:
def add(a,b):
    return a+b

print(add(2, 3)) # add는 정의 되어 있지 않아 오류 발생

5


* **함수의 정의**
  + 정의 시 최초에 def 키워드 사용
  + argument 정의 (함수에 입력으로 전달하는 값을 의미, argument 또는 parameter라고 함) 
  + : (콜론) -> 함수 역시 코드 블록이기 때문에 콜론(:) 필요
  + body (함수의 구현 부분, 함수 역시 코드 블록이기 때문에 들여쓰기 된 부분까지 함수의 코드블록으로 인지 함)
    - 함수를 호출한 코드 (caller)로 함수가 해당 기능을 수행하고 완료된 값(output)을 전달하기 위해 return 키워드 사용
    - 즉, return 이후에 오는 값을 caller로 전달
  + 함수의 네이밍 역시 중요
    - 즉, 어떤 기능을 하는 함수인지 이름으로 최대한 나타날 수 있게 해야함
    - e.g) get_a (x) get_student_name (o)

In [ ]:
# add 함수의 정의
def add(num1, num2):
    # 전달받은 두개의 숫자를 합하여 return 
    return num1 + num2

add(1,3) ## 함수 호출 

In [ ]:
# add2 함수의 정의
def add2(num1, num2):
    print('function is called...')
    print(num1, num2)
    
    # 전달받은 두개의 숫자를 합하여 return 
    return num1 + num2


add2(4,5)

* **exercise** 함수를 정의해 봅시다.
1. 전달받은 3개의 숫자를 한줄에 출력 한뒤, 세 숫자의 곱을 반환하는 함수를 만드세요
2. 문자열 2개를 전달받아 앞의 문자열이 뒤의 문자열에 포함되는지 판단(True, False)하는 함수를 만드세요
3. 정수 한개를 입력 받아, 그것이 소수인지 아닌지 판단(True, False)하는 함수를 만드세요

In [ ]:
def multi(a,b,c):
    result = a*b*c
    return result 
multi(1,3,5)

In [1]:
def intersection (str1,str2):
    if str1 in str2:
        return True 
    return False 

str1 = 'This is python'
str2 = 'python makes me crazy'

intersection (str1,str2)

False

In [5]:
## 소수 찾기 

def prime(num):
    for div in range (2,num):
        if num %div == 0:
            return False 
    return True     


prime(90)

False

In [18]:
##  소인수 분해 

num = 1493
div = 2

a =[]
while num >= div:
    if num % div == 0:
        a.append(div)
        num /= div
        div += 1
    else:
        num -= 1
list(a)          
    

[2, 3, 4, 5, 6]

* **함수의 사용(호출)**
 + 함수명(파라미터1, 파라미터2, ... 파라미터n)
 + 위와 같이 정의 된 함수의 이름과 전달되는 parameter(인자)를 괄호안에 전달하여 함수를 호출
 + 함수가 호출되면 실행의 흐름이 호출자(caller)에서 함수(callee)로 변경 됨
 
 + 함수의 입력(인풋) 파라미터(parameter), 아규먼트(argument)라고도 함

In [ ]:
multi(8,10,18)

In [ ]:
prime(17)

* **함수 네이밍(naming)**
  * 함수 이름으로부터 기능이 명시 
  * 의미와 반대되거나 맞지 않는 이름은 사용금지

In [ ]:
def substract(num1, num2):
    return num1 - num2

def substract(num1, num2):
    return num1 + num2

print(substract(5, 4))

* **parameter(argument) (인자)**
 + 함수에 전달되는 입력(input)
 + 입력이 필요하지 않을 수도, 1개의 입력만 있을 수도, 여러개의 입력이 존재할 수 도 있음 예 프린트 함수 
 + 파라미터로 int, string, float, boolm, list, dict 등등 어떤 파이썬 객체도 전달 가능
 + 심지어, 함수도 함수의 파라미터로 전달 가능
 
 + python의 경우, 타입 명시가 없기 때문에, 함수 생성 시, 의도된 파라미터의 타입에 맞게 입력을 전달하는 것이 중요
 + 또한 파라미터를 전달 할 때, 정의된 순서에 따라 값을 전달하는 것이 중요


In [ ]:
# string이라는 파라미터 1개를 입력으로 받도록 의도
# 또한, 길이를 구하기 위해 의도 하였으므로 리스트, 문자열 등의 타입이 전달 되는 것을 의도
def len2(string):
    return len(string)

print(len2('test'))
print(len2([1, 2, 3, 4]))

* **exercise**)
아래 두 코드는 실행시 오류가 발생합니다. 각각 이유를 말하시오

In [ ]:
# 함수를 잘못 사용한 예
print(len2('test', '45')) # prameter가 하나 들어가도록 의도 되었으나 2개 들어감 
print(len2(23)) ## 문자가 들어가야하는데 숫자가 들어감 

* **Default parameter (기본 인자) ** 
 + 함수의 파라미터에 기본값 지정 가능
 + 파라미터를 명시하지 않을 경우, 지정된 기본값으로 대체

In [ ]:
def print_hello(nums='hello world'): ## nums 는 파라미터  그리고 기본값은 'hello world' 
    print(nums)
    
print_hello('hi world')
# 전달하지 않은 경우 기본 파라미터가 전달
print_hello() ## 기본값으로 출력 

* **기본 파라미터의 다른 예**
 - print 함수
   - sep, end, file등 여러 기본 파라미터를 가짐 

In [ ]:
a = 10
b = 10

# 두 변수 사이의 구분자 재지정
print(a, b, sep='???', end = 'haha')

In [ ]:
# 출력 위치 설정( 기본값 : 표준 출력)
print(a, file=open('a.txt', 'w'))

* **exercise** 
  * 다음 조건을 만족하는 add3함수를 작성하시오.
    * 파라미터 : 정수 파라미터 2개 (두번째 정수가, 입력되지 않은 경우 기본값으로 10 설정)
    * 반환 값 : 입력된 두 정수의 합
  
  

In [ ]:
def add3(a,b=10): ## b 가 호촐되지 않으면 기본값 10으로 돌려줌
    
    return (a+b)

add3(3) ## 파라미터가 비어있으면 기본값이 설정되지 않은 a 때문에 에러남 


* ** Default parameter 사용 시 주의 점**
  * 디폴트 파라미터 뒤에 일반 파라미터가 위치할 수 없음
  
  * e.g) 올바른 예
    > def test(a, b, c = 1)
    
    > def test(a, b = 1, c = 2)
    
    > def test(a = 1, b = 1, c = 3)
    
  * e.g) 올바르지 않은 예
    > def test(a, b = 1, c)
    
    > def test(a = 1, b, c)
    
    > def test(a = 1, b = 1, c)

In [ ]:
def increment_by(a, b = 1):
    return a + b

increment_by(5)

In [ ]:
# 올바른 호출
print(increment_by(45))
print(increment_by(34, 10))

In [ ]:
# 올바르지 않은 호출
print(increment_by()) 

# 이유는? ## a의 인자가 비어있기 때문에 오류 

* **keyword parameter (키워드 파라미터)**
  * 파이썬의 경우, 파라미터에 값을 전달 할 때, 파라미터의 이름을 명시하여 전달 가능
  * 파라미터 이름을 사용하지 않을 경우, 기본적으로 순서에 맞게 전달

In [1]:
def func_test(a, b, c = 1, d = 2):
    print(a, b, c, d)

# 일반적인 호출
func_test(1, 1)
func_test(1, 1, 3, 4)

1 1 1 2
1 1 3 4


In [2]:
# 키워드 파라미터를 사용하는 호출
func_test(b = 2, a = 4)
func_test(1, 7, c = 10, d = 20)
func_test(a = 1, b = 2, c = 4, d = 5)
func_test(c = 1, d = 2, a = 3, b = 2)
func_test(c = 1, a = 3, b = 2)
func_test(d = 10, a = 3, b = 2)

4 2 1 2
1 7 10 20
1 2 4 5
3 2 1 2
3 2 1 2
3 2 1 10


In [3]:
# 잘못된 호출
#func_test(c = 1, d = 2, 1, 3) ## 맨뒤의 숫자들은 순서가 정해지지 않음 
func_test(c = 1, d = 2)
func_test(1, c = 1, d = 2)
func_test()

TypeError: func_test() missing 2 required positional arguments: 'a' and 'b'

* ** 연습문제) range 함수의 구현**
  - loop를 이용하여 range함수를 직접 구현해보세요 

In [ ]:
list(range(10))
list(range(1, 10))
list(range(1, 10, 2))

* **return (리턴)**
 + 기본적으로 함수의 종료를 명시
   + return 옆에 값이나 수식이 있다면 해당 값을 호출자(caller)에게 반환(전달) 
   + return 만 존재하면 None 반환
   + return이 없는 경우, 기본적으로 함수 코드 블록이 종료되면 종료로 간주. 이때도 None 반환

In [ ]:
# 1을 반환하는 함수 생성
def simple():

    print('hahah')
    print('hohohs')
    return 1              ## 리턴값으로 돌려주고 끝나서 다음 프린트 명령은 출력되지 않음 (def 종속에 걸려있는 경우 )

    print('hahah') 

In [ ]:
# simple 함수 호출
simple() ## 마지막 hahaha는 안나옴 

In [ ]:
# simple 함수 호출 결과를 s 변수에 대입
# caller는 현재 cell이며, simple의 코드가 수행된 후, 반환값으로 1을 전달 하였기 때문에
# s에 1이 대입
s = simple()
print(s)

In [ ]:
# 반환값이 없는 함수
def print_something(a): ## a 의 파라미터에는 무조건 조건이 들어가야함 
    print('hi there')
    print(a)
    print('see you again')

In [ ]:
print_something('')

In [ ]:
# 해당 함수는 아무것도 반환하지 않으므로, None 반환
# 즉, 변수 c에는 None 대입
c = print_something('dds')


## print(c)를 출력하라는것은 c의 리턴값을 찾는것 그러나 없으므로 none반환 
print(c) 
c ## 만약 c만 입력한다면 아까 print_something('dds') 하는 명령만 수행


In [4]:
# return 만 명시하여도 None을 반환
def just_return():
    a = 1
    b = 1
    return 
    #return None

c = just_return()
print(c)

None


* **exercise** 
0에서 부터 주어진 정수값의 두배 크기까지의 숫자를 차례대로 갖는 리스트를 생성하여 반환하는 함수를 만드시오

In [5]:
def double(num):
    return(list(range(0,num*2+1)))

In [7]:
double(5)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

* **multiple return (복수 값 반환)**
 + tuple반환을 하여 복수개의 값 리턴 가능

In [8]:
# 이런 경우는 지양하는 것이 좋음
# 함수 한개가 한개의 기능만을 수행하도록 함
def add_sub(a, b):
    return a+b, a-b

def add(a, b):
    return a + b

def sub(a, b):
    return a - b

In [9]:
# 튜플을 반환
c = add_sub(4, 5)

# 튜를 언박싱으로 각각 값을 변수에 대입
e, f = add_sub(4, 5)


print(type(c), c)
print(e, f)



<class 'tuple'> (9, -1)
9 -1


* **variable scope (변수의 범위)** 
 + 변수가 참조 가능한 코드상의 범위를 명시
 + 함수내의 변수는 자신이 속한 코드 블록이 종료되면 소멸됨
 + 이렇게 특정 코드 블록에서 선언된 변수를 **지역변수(local variable)** 이라고 함
 + 반대로 가장 상단에서 정의되어 프로그램 종료 전까지 유지되는 변수를 **전역변수(global variable)**이라고 함
 + 같은 이름의 지역변수와 전역변수가 존재할 경우, 지역변수의 우선순위가 더 높음

In [ ]:
def f(): 
    s = 'java'
    print(s) # 지역변수 s를 출력
    
s = "Python" ## 이건 글로벌 변수 

f()

print(s)

In [ ]:
del s # 아래 예제를 위해 전역 변수 s 삭제

In [ ]:
def f(): 
    s = "Perl" # 로컬 변수 생성
    print(s) 

print(f()) ## 벌써 이미 f() 실행이 되었고 없어졌음 그래서 none 값이 나옴 

print(s) ## 글로벌 함수라서 이미 글로벌 함수로 정의된것을 찾아 출력함 
# s는 함수f의 지역변수였기 때문에 함수가 종료되면서 scope이 종료되어 사라짐, 따라서 오류 발생

In [ ]:
def f():
    print(s) 
    
s = 'Fish'

f()

In [ ]:
def f(): ## 이것만 단독으로 실행하면 값은 안나옴, f() 전체 종속절 print 까지 다 안나옴 
    
    print(s)
    
    global sound 
     ## global 조건은 무조건 s 이전에 나와야 한다 , 하기에 나오는 s는 이제 글로벌이다 라고 선언 , 만약 s보다 global s가 뒤에 있으면 에러:name 's' is assigned to before global declaration
    sound = "dog" # 지역변수가 아닌, 전역변수가 인식 
    
    print(sound)  

    
sound = "cat" 
print(s) # 위의 s = fish 값 대입됨 

print(sound) ## cat 값
f() ##f()에서 호출된 값 


print(sound) ## 글로벌 sound 인 dog 값
print(sound) ## 글로벌 sound 인 dog 값


* 연습문제 
 1. 섭씨 온도를 입력받아, 화씨 온도로 변환하는 함수를 만들고, 섭씨 36.5의 화씨 값을 출력하세요. 화씨 = 섭씨x1.8 + 32
 2. 숫자로 이루어진 리스트를 입력 받아, 그 수 중 홀수인 경우만 합하여 반환하는 함수를 만드세요.
 3. 숫자로 이루어진 리스트를 입력 받아, 그 수 중 가장 큰 수를 반환하는 함수를 만드세요.
 4. 문자열을 입력받아 거꾸로된 문자열을 반환하는 함수를 만드세요. 
 5. 숫자를 입력 받아 factorial을 반환하는 함수를 만드세요. 5! = factorial(5) = 5x4x3x2x1
 6. 숫자로 이루어진 리스트를 입력 받아, 중복을 제거한 리스트만 반환하는 함수를 만드세요. 반환되는 리스트의 순서는 관계 없음 [1,1, 2, 3, 3, 2, 4] -> [1, 2, 3, 4]
 7. 숫자로 이루어진 리스트를 입력 받아, 특정 숫자를 검색하는 함수를 만드세요. 없는 경우 -1 반환
 8. 정렬된 숫자로 이루어진 리스트를 입력 받아, 이진탐색으로 특정 숫자를 검색하는 함수를 만드세요. 없는 경우 -1 반환
 9. 두개의 정수 리스트를 입력 받아, 두 리스트에 모두 포함된 숫자만 갖는 리스트를 반환하시오. 이때, 생성되는 리스트는 정렬할 필요가 없습니다.

In [102]:

def Fe(c):
    return c*1.8 +32 

Fe(36.5)

97.7

In [9]:
import random

nums = random.sample(range(40), 22)

nums


[18,
 11,
 22,
 32,
 2,
 36,
 4,
 25,
 15,
 13,
 38,
 16,
 23,
 27,
 37,
 0,
 39,
 10,
 33,
 8,
 14,
 30]

In [10]:
## 숫자로 이루어진 리스트를 입력 받아, 그 수 중 홀수인 경우만 합하여 반환하는 함수를 만드세요.
def sum_thing(nums):
 
    summ = 0
    for i in nums:
        
        if i & 1: ##??????  i 가 홀수이면 
            summ += i
    return summ

sum_thing(nums)              

223

In [23]:
sum(list(range(1,11)))## practice 

55

In [30]:
def sum_all(lists): ## my practice 
    sum = 0
    for i in lists:
        if i & 1: ## bitwise 이진법 셈, 만약  이진법으로 변형된 i의 마지막자리가 1 과 교집합이 true  라면 :
            sum += i 
    return sum 

print(list1) 
sum_all(list1)


[1, 2, 3, 4, 5, 6]


9

In [97]:
## 13이진법 = 0000 1101 
## 60 이진법 = 0011 1100 
## 13이진법 과 30이진법의 교집합은 a&b = 0000 1100
## https://www.tutorialspoint.com/python3/python_basic_operators.htm

'''Bitwise operator works on bits and performs bit-by-bit operation. Assume if a = 60; and b = 13; Now in binary format they will be as follows −

a = 0011 1100

b = 0000 1101

-----------------

a&b = 0000 1100

a|b = 0011 1101

a^b = 0011 0001

~a = 1100 0011 '''

NameError: name 'binary' is not defined

In [19]:
##문자열을 입력받아 거꾸로된 문자열을 반환하는 함수를 만드세요 (리버싱) - 내답
a = 'merry christmas'


def reverse(n):
    c = sorted(n,reverse = True)
    return c

reverse(a)


['y', 't', 's', 's', 'r', 'r', 'r', 'm', 'm', 'i', 'h', 'e', 'c', 'a', ' ']

In [13]:
def reverse_str(str1):
    return str1[:: -1]

reverse_str(a)

'samtsirhc yrrem'

In [71]:
##숫자를 입력 받아 factorial을 반환하는 함수를 만드세요. 5! = factorial(5) = 5x4x3x2x1
## 재귀함수
def fac (n):
    if n <= 1:
        return 1
    return n*fac(n-1)

fac(5)
fac(4)


24

In [34]:
## 숫자로 이루어진 리스트를 입력 받아, 중복을 제거한 리스트만 반환하는 함수를 만드세요. 반환되는 리스트의 순서는 관계 없음 [1,1, 2, 3, 3, 2, 4] -> [1, 2, 3, 4]

a = {1,1, 2, 3, 3, 2, 4} ## 딕셔너리를 셋으로 만들면 중복제거됨 (셋은 집합과 같음)

b = [1,1,2,3,3,4,2] ## 본래 리스트에서는 리스트로 만들어도 중복제거안됨
def sorting_list(n):
    return set(n)


print(sorting_list(a),sorting_list(b))
print(sorted(a,reverse=True))

{1, 2, 3, 4} {1, 2, 3, 4}
[4, 3, 2, 1]


In [20]:
def remove_d(nums):
    return set(nums) ## set은 중복되는 모든 수를 없애준다 

remove_d(a)

{' ', 'a', 'c', 'e', 'h', 'i', 'm', 'r', 's', 't', 'y'}

In [21]:
## 숫자로 이루어진 리스트를 입력 받아, 특정 숫자를 검색하는 함수를 만드세요. 없는 경우 -1 반환

nums = random.sample(range(35),18)

nums

[33, 9, 5, 27, 32, 29, 20, 25, 18, 19, 15, 34, 0, 13, 24, 26, 22, 31]

In [23]:
def search_somenum (nums, key):
    if key in nums:
        print(key)
    else:
        return -1

search_somenum (nums, 19)

19


In [35]:
## 두개의 정수 리스트를 입력 받아, 두 리스트에 모두 포함된 숫자만 갖는 리스트를 반환하시오. 이때, 생성되는 리스트는 정렬할 필요가 없습니다.
list1 = [4, 7, 15, 35, 78, 25]
list2 = [35, 70, 78, 10, 3, 6]

def items_lists(num1, num2): ## 정수 2 개만 받음 
    result = [] ## 리절트 리스트에 넣어라 
    for i in num1:
        if i in num2:
            result.append(i)
    return result 

items_lists(list1, list2)


[35, 78]

In [26]:
## 두개의 정수 리스트를 입력 받아, 두 리스트에 모두 포함된 숫자만 갖는 리스트를 반환하시오. 이때, 생성되는 리스트는 정렬할 필요가 없습니다.
def items_lists(num1, num2): ## 정수 2 개만 받음 
    result = [] ## 리절트 리스트에 넣어라 
    
    d = {n:1 for n in num2} ## comprehension, 딕셔너리 num2에 있는 키값들의 벨류값은 모두 1씩 
    
    for i in num1:
        if i in d: ## d 의 키값인 i가 있다면 result 리스트로 반환 
            result.append(i)
    return result 

items_lists (list1, list2)


[35, 78, 35]

In [40]:
d = {n:2 for n in list2} ## Comprehension
d


{3: 2, 6: 2, 10: 2, 35: 2, 70: 2, 78: 2}

In [45]:
f= [i**2 for i in range (1,6)]


g= {i:i**2 for i in range (1,6)}
g

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

In [69]:
fruits = {'apple','mango','potatoes','tomatoes'}
juice = {'apple juice', 'mango juice', 'lemon juice','pear juice'}
buckets = {i:len(i) for i in fruits}
list(buckets.items()) ## 튜플의 리스트 형태로 전환 
list(buckets.keys())
for k,v  in zip (fruits, juice):
    print(k,v)


apple pear juice
tomatoes lemon juice
potatoes mango juice
mango apple juice


+ **variable length argument (가변길이 인자)**
 - 전달되는 파라미터의 개수가 고정적이지 않은 경우 사용
 - e.g)
   - print 함수
   - format 함수
  
>***args**,  ****kwargs**

> ***args**    : 파라미터를 튜플의 형태로 전달

> ****kwargs** : 파리미터를 딕셔너리 형태로 전달(네임드 파라미터)

In [77]:
# 파라미터 이름 앞에 *가 붙어 가변인수로 정의
def test_var_args(*args): ## 파라미터가 여러개이다, 몇개인지 모른다 라고 정해줌 
    print("formal arg:", args, type(args))
    
    for arg in args:
        print("another arg:", arg)

test_var_args(1, "two", 'three')
print('-' * 50)

test_var_args(1, 'two')
print('-' * 50)

test_var_args(1)
print('-' * 50)

test_var_args()
print('-' * 50)

formal arg: (1, 'two', 'three') <class 'tuple'>
another arg: 1
another arg: two
another arg: three
--------------------------------------------------
formal arg: (1, 'two') <class 'tuple'>
another arg: 1
another arg: two
--------------------------------------------------
formal arg: (1,) <class 'tuple'>
another arg: 1
--------------------------------------------------
formal arg: () <class 'tuple'>
--------------------------------------------------


In [121]:
# 파라미터 이름 앞에 *가 붙어 가변인수로 정의
def test_var_args(*mmm): ## 파라미터가 여러개이다, 몇개인지 모른다 라고 정해줌 
    print("formal arg:", mmm, type(mmm)) ## mmm의 리스트에 대한 값 출력 
    
    for i in mmm:
        print("another arg:", i)

test_var_args(1, "two", 'three')
print('-' * 50)

test_var_args(1, 'two')
print('-' * 50)

test_var_args(1)
print('-' * 50)

test_var_args()
print('-' * 50)

formal arg: (1, 'two', 'three') <class 'tuple'>
another arg: 1
another arg: two
another arg: three
--------------------------------------------------
formal arg: (1, 'two') <class 'tuple'>
another arg: 1
another arg: two
--------------------------------------------------
formal arg: (1,) <class 'tuple'>
another arg: 1
--------------------------------------------------
formal arg: () <class 'tuple'>
--------------------------------------------------


In [122]:
# 파라미터 이름 앞에 *가 붙어 가변인수로 정의
# 여기서는 첫번째 파라미터는 일반 파라미터
def test_var_args(farg, *args):
    print("formal arg:", args, type(args))
    
    print(farg, end=' ')
    for i in args:
        print("another arg:", i)

test_var_args(1, "two", 'three')
print('-' * 50)

test_var_args(1, 'two')
print('-' * 50)

test_var_args(1)
print('-' * 50)

# 에러 발생. 이유는? ## 에러는 일반 파라미터 1개를 넣으라고 default해줬는데 1개도 들어가 있지 않아서 
test_var_args() ## test_var_args() missing 1 required positional argument: 'farg'
print('-' * 50)

formal arg: ('two', 'three') <class 'tuple'>
1 another arg: two
another arg: three
--------------------------------------------------
formal arg: ('two',) <class 'tuple'>
1 another arg: two
--------------------------------------------------
formal arg: () <class 'tuple'>
1 --------------------------------------------------


TypeError: test_var_args() missing 1 required positional argument: 'farg'

In [123]:
# 위 함수와의 차이점은?
# 두개의 파라미터 모두 일반 파라미터 이기 때문에 2개이 인자만 전달 할 수 있음
def test_tuple_args(farg, tuple_args):
    print("formal arg:", farg)
    print(type(tuple_args))
    for arg in tuple_args:
        print("another arg:", arg)

#test_tuple_args(1, "two", 3) 인자의 개수가 맞지 않아 에러 발생

test_tuple_args(1, ("two", 3))
print('-' * 30)
test_tuple_args(1, "two", 3) 
#test_tuple_args(1) 인자의 개수가 맞지 않아 에러 발생

formal arg: 1
<class 'tuple'>
another arg: two
another arg: 3
------------------------------


TypeError: test_tuple_args() takes 2 positional arguments but 3 were given

* **keyword parameter (키워드 파라미터)**
 - \**가 붙은 경우에는 키워드 파라미터로 인식
 - 즉 함수 호출 시, 파리미터의 이름과 값을 함께 전달 가능

In [31]:
# 키워드 파라미터, 딕셔너리 타입을 가짐
# 키워드 파라미터 형태로 전달해야함
def test_var_kwargs(**kwargs):
    print(type(kwargs))
    for key in kwargs:
        print("another keyword arg: %s: %s"\
        % (key, kwargs[key]))  ##  %s 의 자리에 키 값과 벨류값을 넣어라 

test_var_kwargs(two = 2, three = '3')
test_var_kwargs(aa =1 ,bb =2)

<class 'dict'>
another keyword arg: two: 2
another keyword arg: three: 3
<class 'dict'>
another keyword arg: aa: 1
another keyword arg: bb: 2


In [125]:
def test_var_kwargs(**kwargs):# 인자를 딕셔너리로 모아줌 
    print(type(kwargs))
    for key in kwargs:
        print("another keyword arg: %s: %s" % (key, kwargs[key]))  ##  %s 의 자리에 키 값과 벨류값을 넣어라 ## 딕셔너리는 그룸[키], 인덱스 안되고 키값만 받음

test_var_kwargs(two = "two", three = 'three') 

<class 'dict'>
another keyword arg: two: two
another keyword arg: three: three


In [131]:
name = ('Tom', 'Mary', 'Sharon')

'Hello {}, how are you?'.format(name)

"Hello ('Tom', 'Mary', 'Sharon'), how are you?"

In [71]:
# 일반적인 경우, 아래와 같이 정의 하여 사용
# kwags가 뒤에 위치하는 이유는?
def test_var_args(*args, **kwargs):
    print("formal arg:", args)
    print(type(args))
    print() ## 빈 형태 출력
    
    for arg in args:
        print("another arg:", arg)
        
    for kw in kwargs:
        print(kw, kwargs[kw]) ## 딕셔너리 그룹[키] 키값만 받음 

test_var_args(1, "two", 'three', a = 'aa', b = 'bb')

formal arg: (1, 'two', 'three')
<class 'tuple'>

another arg: 1
another arg: two
another arg: three
a aa
b bb


In [74]:
## 시험문제

def var_args(f_arg, *argv, **kwargs): 
    print("second:", f_arg) ## 일반 파라미터 1개 = f_arg 그래서 파라미터 첫번째 9.14 값이 출력
    
    for kw in kwargs:
        print("Third:", kwargs[kw]) ## 키워드 파라미터의 벨류값'shpark'가 나온다 
    
    for arg in argv: ## 그룹 파라미터의 arg첫번째인 'jepark가 나온다 
        print("first:", arg)
        
var_args(9.14, 'jepark', 100, 'jkpark', 4.5324, keyword = 'shpark')

second: 9.14
Third: shpark
first: jepark
first: 100
first: jkpark
first: 4.5324


* 고정길이 함수에 \*, \** 사용 
  - 튜플 혹은 딕셔너리를 가변길이 인자로 변경하여 전달

In [135]:
def test_var_args_call(arg1, arg2, arg3):
    print("arg1:", arg1)
    print("arg2:", arg2)
    print("arg3:", arg3)

args = ("two", 3, 1)
test_var_args_call(*args) ## 호출할때 사용도 가능 ( 뒷 def 함수의 튜플을 함수 인자로 바로 콜 )

arg1: two
arg2: 3
arg3: 1


In [139]:
def test_var_args_call(arg1, arg2, arg3): ## 함수의 parameter의 개수는 정해져있음 
    print("arg1:", arg1)
    print("arg2:", arg2)
    print("arg3:", arg3)

kwargs = {"arg3": 3, "arg2": "two"}
test_var_args_call(1,**kwargs) ## 호출하는 부분에 씀

arg1: 1
arg2: two
arg3: 3


* e.g) 가변길이 함수의 대표적인 예 # 진짜 중요
 **문자열 포맷 함수**
 - 여러가지 값과 포맷을 이용하여 문자열을 정의할 수 있는 함수
 - {} placeholder를 문자열 내에 위치 시킨 후, 해당 위치에 format함수로 전달된 값으로 대체하여 문자열 생성
 - 포맷 구성은 다음 링크 참조 : https://pyformat.info/

In [58]:
str1 = 'Hi my name is {} and the weather is {}' \
                        .format('younghyo', 'wonderful') ## {} = place holder 개수만큼 들어감 , 가변길이 함수 
print(str1)

Hi my name is younghyo and the weather is wonderful


In [59]:
aa = 'hello {} world {}'.format('mike', 'pop')
print(aa)

a = 'hello world {}'.format('aaron')
print(a)
 
b = 'Hello {}! my name is {} {}'.format(100, 'aaron', 200) ## 값 type 상관 없음 
print(b)



hello mike world pop
hello world aaron
Hello 100! my name is aaron 200


In [60]:
# 순서 명시 가능
# 순서가 명시 되지 않으면 왼쪽부터 0부터 시작
b = 'Hello {1}! my name is {0}'.format('bryan', 'aaron')
print(b)

c = 'Hello {aa}! my name is {bb}'.format(aa = 'Tom', bb = 'Boby')
print(c)

b = 'Hello {1}! my name is {1} {0}'.format(100, 'aaron', 200) ## 값 type 상관 없음 , 슌소룰 바꾸거나 여러번 쓸 수 있음
print(b)

Hello aaron! my name is bryan
Hello Tom! my name is Boby
Hello aaron! my name is aaron 100


In [145]:
# 정수 콤마로 표시 천단위 콤마끊음 
d = 'Wow it is a big number! {:,}'.format(2 ** 30)
print(d)

# 숫자 precision 5자리로 제한
d = 'Wow it is a big number! {:.5}'.format(3.14159262840725) ## 실수에 해당할경우 처음 숫자부터 5개까지 보여주고 자름 
print(d)

# .이후의 precision 5자리로 제한
d = 'Wow it is a big number! {:.5f}'.format(3.14159262840725) ## 소수점 이후 5자리만 보존 하겠다 

print(d)



Wow it is a big number! 1,073,741,824
Wow it is a big number! 3.1416
Wow it is a big number! 3.14159


In [61]:
# 최소 7자리는 필요 .이후 precision 5자리로 제한
d = 'Wow it is a big number! {:7.5f}'.format(1213.14159262840725)
print(d)

# 최소 7자리는 필요 .이후 precision 1자리로 제한
# 숫자의 경우 오른쪽 부터 채워짐. (패딩이 왼쪽에 위치)
d = 'Wow it is a big number! {:7.1f}'.format(1213.14159262840725)
print(d)

# 패딩을 넣은 경우 = 0 
d = 'Wow it is a big number! {:07.1f}'.format(1213.14159262840725)
print(d)

Wow it is a big number! 1213.14159
Wow it is a big number!  1213.1
Wow it is a big number! 01213.1


In [146]:
# old style 문자열 포매팅
d = 'Hello %s %d %s' % ('aaron', 4454, 'cathy') ## 튜플로 받음 
print(d)

Hello aaron 4454 cathy


In [78]:
old_style = ' hi nice to meet you, %s , %d %% humid today, so it is %s now' %('Mark', 60,'hot')

print(old_style) ## 프린트 함수는 스트링이 아닌 값만 출력해줌 

 hi nice to meet you, Mark , 60 % humid today, so it is hot now


In [147]:
print('{}'.format(420))
print('{0}'.format(420))

# : 뒤에 포매팅 문법이 위치
print('{:4}'.format(420)) # 4만큼의 길이 필요
print('{:04}'.format(420)) # 4만큼의 길이가 필요하고 없으면 빈 공간은 0으로 채움
print('{:06}'.format(4200)) # 6만큼의 길이가 필요하고 없으면 빈 공간은 0으로 채움
print('{:4,}'.format(4200))
print('{:4,}'.format(42000))
print('{:6.3f}'.format(3.14159223232)) ## 소수점포함하여 6칸을 채우고 실수자리 3자리만 표시해라 (라운딩하여라)
print('{:06.3f}'.format(3.14159223232))## 소수점 포함해서 6칸을 채우되 빈칸이 있으면 0으로 채우고, 실수자리 3자리표시해라 (라운딩해라)

420
420
 420
0420
004200
4,200
42,000
 3.142
03.142


* **1st class citizen(일등 시민)**
 + 프로그래밍 언어에서 1st class라는 의미는 해당 객체를 사용하는데 있어서 아무런 **"제약없음"**을 의미 
 # 함수 역시 객체 그래서 변수에 대입 가능 호출 가능 
 + python에서는 모든 것이 객체(object)!
 + 숫자, 문자열, 튜플, 리스트, 딕셔너리 등등, 존재하는 모든 값을 객체로 볼 수 있음
 + 함수 역시 객체이며, 객체이기 때문에 변수의 값에 할당하거나, 함수의 파라미터로 전달하거나, 리스트의 원소가 되는 등 일반 타입처럼 사용 할 수 있음

In [ ]:
def a():
    return 1
def b():
    return a()
def c():
    return b() ## 이렇게 하면 되긴 하지만 꼬임, stack 알고리즘 이라서 많이 쌓이다가 오류남 

In [148]:
def test1():
    print(23)
    
def run_something(func):
    func()
    
test2 = test1 # 함수를 변수에 대입
test2()

print(test1, type(test1))
run_something(test1)

23
<function test1 at 0x0000023EA3770268> <class 'function'>
23


## Recursive Function

* 재귀함수
* 내부적으로 자기 자신을 호출함 (러시아 인형, 혹은 프랙탈을 연상)
* 점화식으로 표현되는 모든 수식을 간결히 표현할 수 있음
* 종료조건이 꼭! 필요함 (명시하지 않으면 무한 호출반복)
* 코드는 간결하나, 호출에 따른 오버헤드가 큼
* 따라서 성능과 코드의 간결함에 대한 trade-off를 고려하여 구현이 필요


In [149]:
# factorial을 다시 작성해 봅시다.
def factorial(n):
    mul = 1
    for i in range(2, n+1):
        mul *= i
        
    return mul

# assert 함수 : 주어진 값이 True가 아니면 에러를 발생
# 주로 간단한 테스트에 사용
assert(factorial(5) == 120)
assert(factorial(4) == 24)
assert(factorial(1) == 1) ## ture 아니면 에러 , assert 함수는 ture 일 경우 아무것도 출력 안하지만 false 면 에러를 출력 
factorial(6)

720

In [152]:
def rf(n):
    if n == 0:
        return 1
    
    return n * rf(n-1)

assert(rf(5) == 120)
assert(rf(4) == 24)
assert(rf(1) == 1)
rf(6)

720

* **%timeit**
 - 해당 코드를 수행하여 평균 수행시간을 계산

In [153]:
%timeit x = range(10000)

The slowest run took 7.28 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 293 ns per loop


In [154]:
%timeit range(10000)

The slowest run took 4.37 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 293 ns per loop


In [155]:
%timeit factorial(10000)

10 loops, best of 3: 28.7 ms per loop


In [ ]:
# 재귀함수 호출의 값이 기본적으로 5000이기 때문에 늘려줌
import sys
sys.setrecursionlimit(200000) ## recursion 반복 되풀이

%timeit rf(10000)

In [1]:
def fibonacci(n):
    fibo = [1, 1]
    for i in range(2, n):
        fibo.append(fibo[i-1] + fibo[i-2])

    return fibo[-1]

print(fibonacci(15))
%timeit fibonacci(20)

610
100000 loops, best of 3: 4.68 µs per loop


In [2]:
def recursive_fibonacci(n):
    if n == 1 or n == 2:
        return 1
    
    return recursive_fibonacci(n-1) + recursive_fibonacci(n-2) ## 재귀호출이 두번 한꺼번에 요청됨 ## 너무 많은 시간 이 필요  2**n 번 걸림 

print(recursive_fibonacci(15))
%timeit recursive_fibonacci(20)

610
100 loops, best of 3: 2.59 ms per loop


* **연습문제**
1. 계단오르기 문제. 계단을 한번에 1,2 혹은 3칸까지 오를 수 있다고 할 때

n층의 계단을 오르기 위해서는 몇 가지 경우가 가능한가?

In [ ]:
## 점화식 = f(3)= f(2)+f(1)
        # f(4)= f(3)+f(2)+f(1)
    
def stairs(n):
    if n ==1 or n == 2:
        return n
    if n == 3 
        return 4
    
    return stairs(n-1)+stairs(n-2)+ stairs(n-3)  ## 너무오래 걸림


In [80]:
memo = {}

## 점화식 = f(3)= f(2)+f(1)
        # f(4)= f(3)+f(2)+f(1)
    
def stairs(n):
    if n ==1 or n == 2:
        return n
    if n == 3: 
        return 4
    if n in memo:
        return memo[n]
    
    memo[n] = stairs(n-1)+stairs(n-2)+ stairs(n-3)## 만약 메모가 없다면 처음부터 다시계산하는데 너무오래 걸림
    return memo[n]
stairs(5)
stairs(21)

memo




{4: 7,
 5: 13,
 6: 24,
 7: 44,
 8: 81,
 9: 149,
 10: 274,
 11: 504,
 12: 927,
 13: 1705,
 14: 3136,
 15: 5768,
 16: 10609,
 17: 19513,
 18: 35890,
 19: 66012,
 20: 121415,
 21: 223317}

* **memozation(메모이제이션)**
  - 이미 계산한 값은 따로 저장해두고, 필요하면 다시 불러와서 사용 하는 기법  - dynamic programming의 재귀적 구현 기법에서 많이 사용

In [96]:

def fibo(n):
    
    if n==1 or n==2:
        return 1
    
    if n>=3:
        return fibo(n-1)+fibo(n-2)
    if n in memo1:
        return memo1[n]
    
fibo(10) 



55

In [16]:
memo = {}
def recursive_fibonacci(n):
    if n == 1 or n == 2:
        return 1
    
    if n in memo:
        return memo[n]
    
    memo[n] = recursive_fibonacci(n-1) + recursive_fibonacci(n-2)
    return memo[n]


print (recursive_fibonacci(13))
%timeit recursive_fibonacci(50)

memo

233
The slowest run took 246.56 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 235 ns per loop


{3: 2,
 4: 3,
 5: 5,
 6: 8,
 7: 13,
 8: 21,
 9: 34,
 10: 55,
 11: 89,
 12: 144,
 13: 233,
 14: 377,
 15: 610,
 16: 987,
 17: 1597,
 18: 2584,
 19: 4181,
 20: 6765,
 21: 10946,
 22: 17711,
 23: 28657,
 24: 46368,
 25: 75025,
 26: 121393,
 27: 196418,
 28: 317811,
 29: 514229,
 30: 832040,
 31: 1346269,
 32: 2178309,
 33: 3524578,
 34: 5702887,
 35: 9227465,
 36: 14930352,
 37: 24157817,
 38: 39088169,
 39: 63245986,
 40: 102334155,
 41: 165580141,
 42: 267914296,
 43: 433494437,
 44: 701408733,
 45: 1134903170,
 46: 1836311903,
 47: 2971215073,
 48: 4807526976,
 49: 7778742049,
 50: 12586269025}

* recursive function 연습문제
 1. 숫자 리스트를 인자로 받아 그 숫자의 합을 구하는 재귀함수를 작성하시오.
 2. 숫자를 입력받아 각 자리수의 합을 구하는 재귀함수를 작성하시오.
 3. 주어진 두 수의 최대 공약수를 찾는 재귀함수를 작성하시오.
    A와 B의 최대 공약수는 B와 A%B의 최대 공약수와 같다.

In [7]:
import random
nums = random.sample (range(30),15)

nums



[16, 17, 20, 23, 21, 26, 10, 22, 24, 15, 4, 2, 12, 18, 25]

In [28]:
a = [ 3, 7, 1, 2 ]

def list_sum(nums, n):
    if n == 0:
        return nums[n]
    return list_sum(nums, n-1) + nums[n]

print(list_sum(a, len(a)-1))

13


In [29]:
## 선생님 답

import random
nums = random.sample(range(1000), 10)

def sum_list(nums, k):
    if k == 0:
        return nums[k]
    return sum_list(nums, k-1) + nums[k]

sum_list(nums, len(nums)-1)

4861

In [ ]:
#nums = random.sample(range(1000), 200)
def sum_list(nums, s, e):
    if s == e:
        return nums[s]
    
    return sum_list(nums, s, (s+e)//2) + sum_list(nums, (s+e)//2 + 1, e)

sum_list(nums, 0, len(nums)-1)

In [18]:
def gcd(A,B):
    if B == 0: 
        return A 
    return gcd (B, A % B) 

In [21]:
gcd(40,300)

20

In [22]:

## 숫자를 입력받아 각 자리수의 합을 구하는 재귀함수를 작성하시오.
def sum_digit(num):
    if num < 10:
        return num
    return sum_digit(num // 10) + num % 10

sum_digit(7321)

13

파이썬(Python)에서  올림함수 반올림함수 내림함수 
 
파이썬(python)에서 소수을 반올림하거나 올림, 혹은 내림(버림)하는 함수 소개.
 
1. 기본적으으로 제공하는 함수
   버림 함수(내림 함수)는 int()함수를 사용한다.
      예) print int(3.1584)     ==> 3이 출력된다.
 
   반올림 함수는 round()함수를 사용한다.
      예) print round(3.56)    ==> 4가 출력된다.
           print round(3.14)    ==> 3이 출력된다.
  
   올림함수는 round()함수를 응용한다.
      예) a = 3.14 일때
           print round(a + 0.5)  ==> 4가 출력된다.
 
           a = 3. 564 일때
           print round(a + 0.5) ==> 4가 출력된다.
 
2. 라이브러리를 사용하기
 
import math
math.floor(3.223)  ==> 3   버림함수
math.ceil(3.223)  ==> 4    올림함수
round(3.223)      ==> 3     반올림함수
[출처] 파이썬(Python) 올림함수 반올림함수 내림함수|작성자 네미건블